File: server/src/models/userModel.ts
import { Schema, model } from "mongoose"
import Validator from "validator"

const userSchema = new Schema(
    {
        username: {
            type: String,
            required: true,
            min: [4, "Name must atleast contain 4 charcaters"],
            max: [20, "Name can only contain 20 charcaters"],
            unique: true,
            trim: true,
            validate: {
                validator: function (name: string) {
                    // Allow only Alphabets,Numbers,Undersores and space
                    return /^[a-z0-9-_\s]+$/i.test(name)
                },
                message: `Username can contain only alphabets,numbers , whitespace and underscores`,
            },
        },
        email: {
            type: String,
            unique: true,
            trim: true,
            lowercase: true,
            required: true,
            validate: [Validator.isEmail, "Please enter a valide Email!"],
        },
        publicId: {
            type: String,
            unique: true,
            required: true,
        },
        profilePic: {
            type: String,
            default:
                "https://res.cloudinary.com/dciwowqk7/image/upload/v1660972618/user/default-profile-pic.jpg",
        },
        hash: {
            type: String,
            required: true,
        },
        salt: {
            type: String,
            required: true,
        },
    },
    { timestamps: true },
)

userSchema.pre("save", function (next) {
    console.log(this)
    next()
})

const User = model("User", userSchema)
export default User

File: app/lib/models/user_model.dart
class PublicUserData {
  final String name;
  final String publicId;
  final String profilePic;
  PublicUserData(
      {required this.name, required this.profilePic, required this.publicId});
}

class PersonalData extends PublicUserData {
  final String email;
  final String token;
  PersonalData(
      {required this.email,
      required this.token,
      required super.name,
      required super.publicId,
      required super.profilePic});

  factory PersonalData.fromJson(Map<String, String> json) {
    return PersonalData(
        email: json['email'] as String,
        token: json['token'] as String,
        name: json['username'] as String,
        publicId: json['publicId'] as String,
        profilePic: json['profilePic'] as String);
  }

  PublicUserData get getPublicData => this;
}
